official/utils/logs/logger.py

```diff
@@ -183,6 +185,65 @@ def log_run_info(self, model_name, dataset_name, run_params):
                            e)
 
 
+class BenchmarkBigQueryLogger(BaseBenchmarkLogger):
+  """Class to log the benchmark information to BigQuery data store."""
+
+  def __init__(self,
+               bigquery_uploader,
+               bigquery_data_set,
+               bigquery_run_table,
+               bigquery_metric_table,
+               run_id):
+    super(BenchmarkBigQueryLogger, self).__init__()
+    self._bigquery_uploader = bigquery_uploader
+    self._bigquery_data_set = bigquery_data_set
+    self._bigquery_run_table = bigquery_run_table
+    self._bigquery_metric_table = bigquery_metric_table
+    self._run_id = run_id
+
+  def log_metric(self, name, value, unit=None, global_step=None, extras=None):
+    """Log the benchmark metric information to bigquery.
+
+    Args:
+      name: string, the name of the metric to log.
+      value: number, the value of the metric. The value will not be logged if it
+        is not a number type.
+      unit: string, the unit of the metric, E.g "image per second".
+      global_step: int, the global_step when the metric is logged.
+      extras: map of string:string, the extra information about the metric.
+    """
+    metric = _process_metric_to_json(name, value, unit, global_step, extras)
+    if metric:
+      # Starting new thread for bigquery upload in case it might take long time
+      # and impact the benchmark and performance measurement.
+      thread.start_new_thread(
+          self._bigquery_uploader.upload_benchmark_metric_json,
+          (self._bigquery_data_set,
+           self._bigquery_metric_table,
+           self._run_id,
+           [metric]))
+
+  def log_run_info(self, model_name, dataset_name, run_params):
+    """Collect most of the TF runtime information for the local env.
+
+    The schema of the run info follows official/benchmark/datastore/schema.
+
+    Args:
+      model_name: string, the name of the model.
+      dataset_name: string, the name of dataset for training and evaluation.
+      run_params: dict, the dictionary of parameters for the run, it could
+        include hyperparameters or other params that are important for the run.
+    """
+    run_info = _gather_run_info(model_name, dataset_name, run_params)
+    # Starting new thread for bigquery upload in case it might take long time
+    # and impact the benchmark and performance measurement.
+    thread.start_new_thread(
```

Comment:
Q1: If we spawn a thread to do this, what happens to errors raised? They will not interrupt the run, correct? What will happen in that scenario? 
Q2: Under normal circumstances, we allow up to num_cpus for input processing, whereas this will require one separate CPU thread. In practice, it's really unlikely that this will matter, but it's important to be aware of possible heisenbug situations, where we are somehow slowing down runs by trying to observe them, especially when we get to CPU benchmarks. Thoughts? I guess we can compare the speed of non-streaming to streaming runs periodically to check for such an effect, but can this be done in a more automated way?

Reply:
1. We currently log the error and does not stop the training process.
2. CPU benchmark might be harder since all the process running on the machine will have the potential impact of the performance. We can setup a separate run for non-streaming benchmark when we start testing on CPU.

Reply:
Okay-- let's just make sure to record here in the comments that users should be aware of this potential performance implication, especially for CPU training, and run tests accordingly.

Reply:
Done.